apps/omg/lib/omg/state/core.ex

```diff
@@ -99,31 +99,24 @@ defmodule OMG.State.Core do
   Recovers the ledger's state from data delivered by the `OMG.DB`
   """
   @spec extract_initial_state(
-          utxos_query_result :: [list({OMG.DB.utxo_pos_db_t(), OMG.Utxo.t()})],
           height_query_result :: non_neg_integer() | :not_found,
           child_block_interval :: pos_integer()
         ) :: {:ok, t()} | {:error, :top_block_number_not_found}
-  def extract_initial_state(utxos_query_result, height_query_result, child_block_interval)
-      when is_list(utxos_query_result) and is_integer(height_query_result) and is_integer(child_block_interval) do
-    # extract height, last deposit height and utxos from query result
-    height = height_query_result + child_block_interval
-
-    state = %__MODULE__{
-      height: height,
-      utxos: UtxoSet.init(utxos_query_result)
-    }
+  def extract_initial_state(height_query_result, child_block_interval)
+      when is_integer(height_query_result) and is_integer(child_block_interval) do
+    state = %__MODULE__{height: height_query_result + child_block_interval}
 
     {:ok, state}
   end
 
-  def extract_initial_state(
-        _utxos_query_result,
-        :not_found,
-        _child_block_interval
-      ) do
+  def extract_initial_state(:not_found, _child_block_interval) do
     {:error, :top_block_number_not_found}
   end
 
+  @spec with_utxos(t(), UtxoSet.query_result_t()) :: t()
+  def with_utxos(%Core{utxos: utxos} = state, utxos_query_result),
```

Comment:
`defp`? also `exec` would have to be `defp`

OTOH let's contrast this approach with the more explicit `def with_utxos`, which enriches the UTXOset then `exec` which remains untouched. This has the benefit of smaller impact on the API & tests.

Reply:
after discussion: public `with_utxos` and public `exec` seem more elegant. 

Just let's treat `exec` and other UTXOSet-operating functions assume that the UTXOSet operated on is well prepared. `with_utxos` prepares it, and must caare about all the edge cases (like the `recently_spent` - case discussed)

Let's try this